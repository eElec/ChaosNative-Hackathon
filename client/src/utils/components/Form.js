import React from 'react';

import styles from '@assets/css/Form.module.css';

const Form = ({
	placeholder,
	handleChange,
	value,
	label,
	PreIcon,
	PostIcon,
	width,
	type
}) => {
	return (
		<div className={styles.form}>
			{label ? <span className={styles.label}>{label}: </span> : <></>}
			{PreIcon ? <PreIcon className={styles.preIcon} /> : <></>}
			<input
				style={{ width: `${width}ch` }}
				type={type?type:'text'}
				value={value}
				placeholder={placeholder}
				onChange={handleChange}
			/>
			{/* Icons for date and time values are not available in Firefox */}
			{PostIcon ? <PostIcon className={styles.postIcon} /> : <></>}
		</div>
	);
};

export default Form;
