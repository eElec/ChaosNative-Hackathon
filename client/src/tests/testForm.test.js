//https://levelup.gitconnected.com/how-to-write-unit-tests-with-react-testing-library-d9624fd2b707
import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import Form from '../utils/components/Form';

const value = 'ABC';

it('matches snapshot', () => {
	const { asFragment } = render(<Form />);

	expect(asFragment()).toMatchSnapshot();
});

describe('Form Mutate', () => {
	const mockChangeValue = jest.fn();
	let values = 'ABC';

	it('mutate form', () => {
		const { getByTestId, rerender } = render(
			<Form value={values} handleChange={mockChangeValue} />
		);

		act(() => {
			fireEvent.change(getByTestId('form-test'), {
				target: { value: 'CDE' },
			});
		});

		rerender(<Form value={values} handleChange={mockChangeValue} />);
		expect(getByTestId('form-test').value).toBe('ABC');
		expect(mockChangeValue).toBeCalledTimes(1);
	});

	function mutateValue(e) {
		values = e.target.value;
	}
	it('actual mutate form', () => {
		const { getByTestId, rerender } = render(
			<Form value={values} handleChange={mutateValue} />
		);

		act(() => {
			fireEvent.change(getByTestId('form-test'), {
				target: { value: 'CDE' },
			});
		});

		rerender(<Form value={values} handleChange={mockChangeValue} />);
		expect(getByTestId('form-test').value).toBe('CDE');
	});
});
