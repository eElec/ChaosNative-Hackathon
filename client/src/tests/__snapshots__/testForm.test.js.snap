// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="form"
  >
    <input
      data-testid="form-test"
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
